Modules/_collectionsmodule.c

```diff
@@ -537,12 +538,24 @@ deque_copy(PyObject *deque)
         Py_DECREF(new_deque);
         return NULL;
     }
-    if (old_deque->maxlen < 0)
-        return PyObject_CallFunctionObjArgs((PyObject *)(Py_TYPE(deque)),
-                                            deque, NULL);
-    else
-        return PyObject_CallFunction((PyObject *)(Py_TYPE(deque)), "Oi",
-            deque, old_deque->maxlen, NULL);
+    if (old_deque->maxlen < 0) {
+        result = PyObject_CallFunctionObjArgs((PyObject *)(Py_TYPE(deque)),
+                                              deque, NULL);
+    }
+    else {
+        result = PyObject_CallFunction((PyObject *)(Py_TYPE(deque)), "Oi",
+                                       deque, old_deque->maxlen, NULL);
+    }
+    if (result != NULL &&
+        !PyObject_IsInstance(result, (PyObject *)&deque_type))
```

Comment:
1) `PyObject_IsInstance()` can raise an exception.
2) Even if `PyObject_IsInstance()` returns 1, the result can have the binary layout different from dequeobject (because `__instancecheck__`), and the following concatenation or repeating will crash. Use `PyObject_TypeCheck()` instead.